crates/ra_project_model/src/sysroot.rs

```diff
@@ -113,6 +116,26 @@ fn try_find_src_path(cargo_toml: &Path) -> Result<PathBuf> {
     Ok(sysroot_path.join("lib/rustlib/src/rust/src"))
 }
 
+fn try_install_rust_src(cargo_toml: &Path) -> Result<PathBuf> {
```

Comment:
Here, we have an implicit invariant that we need to check if `try_find_src_path` works before calling `install`, otherwise it'll be called twice. Let's instead hide this mutating bit in a single idempotent function function which installs only if necessary:

```rust
fn get_or_install_rust_src(cargo_toml: &Path) -> Result<PathBuf> {
    
}
```

The benefit here is that, but not exposing non-idempotent interface, you completely prevent missuse.

In this particular case this is nitpicking and not really necessary, but it's a good idea to just *always* provide `ensure_x` instead of `check_x / do_x` pairs, so that no extra thinking is necessary.

Reply:
I'm tempted to define the `get_or_install_rust_src` function and move both old functions into the body of the new one, unchanged:

```rust
fn get_or_install_rust_src(cargo_toml: &Path) -> Result<PathBuf> {
    fn try_find_src_path(cargo_toml: &Path) -> Result<PathBuf> {
        // old body, wall of text
    }
    fn try_install_rust_src(cargo_toml: &Path) -> Result<PathBuf> {
        // old body, wall of text
    }

    try_find_src_path(cargo_toml)?.or_else(|| try_install_rust_src(cargo_toml))
}
```

Would that suffice?

Reply:
yup

Reply:
Great, done.
I could have extracted the common `if !rustc_output.status.success() {` checks into a separate method either, but maybe it's a bit too much for two repetitions only.